### PR TITLE
ESP32: Add OTARequstor to all-clusters-app

### PR DESCRIPTION
**Pull request by wqx6**
#### Problem
Add ota commands to chip shell for ESP32 platform in all-clusters-app

#### Change overview
1. Enable OTAProvider client cluster in all-clusters-app
2. Add ota commands to chip shell for esp32 platform in all-clusters-app

#### Testing
Test on ESP32C3 with a Linux OTAProvider
1. Input 'matter ota query 1 {Provider-NodeId} 0 {Provider-IPAddr} 5540' to query the image
2. Apply the image and reboot, then the new image will start to run.

### CODE REVIEW

File: src/lib/shell/commands/Ota.cpp [hunk 1]
```diff
@@ -167,7 +169,16 @@ void OnQueryImageFailure(void * /* context */, EmberAfStatus status)
     ChipLogError(SoftwareUpdate, "QueryImage failed: %" PRIu16, static_cast<uint16_t>(status));
 }
 
-void OnQueryImageConnection(void * /* context */, DeviceProxy * deviceProxy)
+} // namespace
+} // namespace Shell
+} // namespace chip
+
+#include <controller-clusters/zap-generated/CHIPClusters.h>
```
Comment (Damian-Nordic) [inline src/lib/shell/commands/Ota.cpp:176]: Why has it been moved here?

**Review by Damian-Nordic — APPROVED** [review]

File: src/lib/shell/commands/Ota.cpp (hunk 1 above)
Comment (wqx6) [reply src/lib/shell/commands/Ota.cpp:176]: I moved the include line here to avoid an error `Include not allowed` when compiling all-clusters-app.

File: examples/all-clusters-app/esp32/main/Kconfig.projbuild [hunk 2]
```diff
@@ -140,3 +140,12 @@ depends on ENABLE_PW_RPC
             about available pin numbers for UART.
 
 endmenu
+
+menu "OTA Options"
+
+    config ENABLE_OTA_REQUESTOR
+        bool "enable ota requestor"
```
Comment (dhrishi) [inline examples/all-clusters-app/esp32/main/Kconfig.projbuild:147]: Enable OTA Requestor

File: examples/all-clusters-app/esp32/main/Kconfig.projbuild [hunk 3]
```diff
@@ -140,3 +140,12 @@ depends on ENABLE_PW_RPC
...
+        default y
+        help
+            Enable this option to enable the ota requestori
```
Comment (dhrishi) [inline examples/all-clusters-app/esp32/main/Kconfig.projbuild:150]: OTA Requestor

File: src/platform/ESP32/InetPlatformConfig.h [hunk 4]
```diff
@@ -41,3 +41,4 @@
 #ifndef INET_CONFIG_NUM_UDP_ENDPOINTS
 #define INET_CONFIG_NUM_UDP_ENDPOINTS CONFIG_NUM_UDP_ENDPOINTS
 #endif // INET_CONFIG_NUM_UDP_ENDPOINTS
+#define INET_CONFIG_ENABLE_IPV4 1
```
Comment (dhrishi) [inline src/platform/ESP32/InetPlatformConfig.h:44]: Why is this required? For the OTA examples?

**Comment by dhrishi** [comment]
@shubhamdp PTAL

**Review by woody-apple — APPROVED** [review]

File: examples/all-clusters-app/esp32/main/Kconfig.projbuild (hunk 2 above)
Comment (wqx6) [reply examples/all-clusters-app/esp32/main/Kconfig.projbuild]: Chaged

File: examples/all-clusters-app/esp32/main/Kconfig.projbuild (hunk 3 above)
Comment (wqx6) [reply examples/all-clusters-app/esp32/main/Kconfig.projbuild]: Done

File: src/platform/ESP32/InetPlatformConfig.h (hunk 4 above)
Comment (wqx6) [reply src/platform/ESP32/InetPlatformConfig.h:44]: I added this to make function `Inet::IPAddress::FromString` work. It is a function transfering an IP String to `Inet::IPAddress`.

File: src/lib/shell/commands/DFUManager_esp32.h [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+};
+
+inline CHIP_ERROR DFUManager::ApplyUpdate()
+{
+    return DeviceLayer::Internal::ESP32Utils::MapError(OTAUpdater::GetInstance().Apply());
+}
+
+inline CHIP_ERROR DFUManager::DiscardUpdate()
+{
+    return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
```
Comment (shubhamdp) [inline src/lib/shell/commands/DFUManager_esp32.h:54]: Is this not similar to `OTAUpdater::Abort()`?

**Review by shubhamdp — APPROVED** [review]

File: src/lib/shell/commands/DFUManager_esp32.h (hunk 5 above)
Comment (wqx6) [reply src/lib/shell/commands/DFUManager_esp32.h:54]: This function is called when there is an `EMBER_ZCL_OTA_APPLY_UPDATE_ACTION_DISCONTINUE` action during applying the update. It is different from `OTAUpdater::Abort()`,

File: src/lib/shell/commands/OTAUpdater_esp32.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+inline esp_err_t OTAUpdater::Write(const void * data, size_t length)
+{
+    if (otaUpdateInProgress == false)
+    {
+        return ESP_ERR_INVALID_STATE;
+    }
+    esp_err_t err = esp_ota_write(otaUpdateHandle, data, length);
+    if (err != ESP_OK)
```
Comment (chshu) [inline src/lib/shell/commands/OTAUpdater_esp32.h:86]: Do we really want to abort here? If so, use ESP_ERROR_CHECK() to simplify the code.

File: src/lib/shell/commands/OTAUpdater_esp32.h [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+inline esp_err_t OTAUpdater::Abort(void)
+{
+    if (otaUpdateInProgress == false)
+    {
+        return ESP_ERR_INVALID_STATE;
+    }
+
+    ESP_LOGI(TAG, "Abort");
```
Comment (chshu) [inline src/lib/shell/commands/OTAUpdater_esp32.h:104]: nit: OTA Abort

File: src/lib/shell/commands/OTAUpdater_esp32.h (hunk 6 above)
Comment (wqx6) [reply src/lib/shell/commands/OTAUpdater_esp32.h:86]: The `Abort()` function here will call `esp_ota_abort` which gets the OTA entry from the OTA handle then removes it from entries list and frees it. It's different from `ESP_ERROR_CHECK()`.

File: src/lib/shell/streamer_esp32.cpp [hunk 8]
```diff
@@ -51,7 +51,10 @@ int streamer_esp32_init(streamer_t * streamer)
     setvbuf(stdin, NULL, _IONBF, 0);
     esp_vfs_dev_uart_port_set_rx_line_endings(CONFIG_ESP_CONSOLE_UART_NUM, ESP_LINE_ENDINGS_CR);
     esp_vfs_dev_uart_port_set_tx_line_endings(CONFIG_ESP_CONSOLE_UART_NUM, ESP_LINE_ENDINGS_CRLF);
-
+    if (uart_is_driver_installed(CONFIG_ESP_CONSOLE_UART_NUM))
```
[the remark this answers is not in the document]
Comment (wqx6) [reply src/lib/shell/streamer_esp32.cpp]: Done

File: src/lib/shell/commands/OTAUpdater_esp32.h (hunk 7 above)
Comment (wqx6) [reply src/lib/shell/commands/OTAUpdater_esp32.h]: Done

**Review by chshu — APPROVED** [review]

**Comment by shubhamdp** [comment]
@wqx6, #12560 is merged, I think we should be implementing Interfaces from there. Should we implement them here or in follow-up PR?

**Review by andy31415 — APPROVED** [review]

**Comment by wqx6** [comment]
> @wqx6, #12560 is merged, I think we should be implementing Interfaces from there. Should we implement them here or in follow-up PR?

Sure, I think using the interfaces form  `src/app/clusters/ota-requestor` is better.

**Comment by bzbarsky-apple** [comment]
@wqx6 Unfortunately, this PR broke OTA Provider server for all-clusters-app due to a limitation in ZAP: In its current form it randomly does not allow the client and server parts of the same cluster to coexist as far as I can tell (@tecimovic).  If you look at `zzz_generated/all-clusters-app/zap-generated/IMClusterCommandHandler.cpp` after this change it looks like this:
```
namespace OtaSoftwareUpdateProvider {

void DispatchServerCommand(CommandHandler * apCommandObj, const ConcreteCommandPath & aCommandPath, TLV::TLVReader & aDataTlv)
{
    ReportCommandUnsupported(apCommandObj, aCommandPath);
}

} // namespace OtaSoftwareUpdateProvider
```

Worse yet, this is not stable: sometimes it ends up including the right code in there and sometimes it does not.  This is causing very common CI failures.  And I am not sure that putting the client on a different endpoint from the server would necessarily help, though maybe it's worth trying.

I'm also a little confused by the fact that the PR title talks about OTARequestor but all the changes are about OTAProvider client....

File: src/platform/ESP32/InetPlatformConfig.h (hunk 4 above)
Comment (andy31415) [reply src/platform/ESP32/InetPlatformConfig.h:44]: This will have to be reverted - you are not  allowed to assume IPV4 is enabled.

Comment (andy31415) [reply src/platform/ESP32/InetPlatformConfig.h:44]: #12700 reverts this. Builds with IPV4 off have to be valid in Matter (SPEC requires IPV6 only and IPV6 is mandatory for thread compatibility)

**Comment by wqx6** [comment]
@bzbarsky-apple I enabled OTA Provider Client because OTA Provider commands `QueryImage` `ApplyUpdateRequest` was used in `src/lib/shell/commands/Ota.cpp`. But in fact, the `OTASoftwareUpdateProviderCluster` used in chip shell was from controller data-model, so the OTAProvider client can be removed in all-clusters-app. I will raise a PR to remove it.

**Comment by bzbarsky-apple** [comment]
> I wonder why the client and the server of the same cluster can't coexist.

Because ZAP is buggy....

**Comment by tima-q** [comment]
@bzbarsky-apple @wqx6 I've put up this PR #13256 last year before the holiday's. It basically removes all stubs from the applications and uses the data model files for all ota-requestor-server implementation.
Side effect was I needed to added the correct provider client/requestor server to all applications (shell, all-clusters-app) that had a stubbed implementation.
Now I see that kind of enabling broke some stuff due to ZAP. How can I proceed to integrate this ? 
End goal is to get the lighting-app 'OTA-ready' - preferably in a clean way eg just enabling the necessary clusters.
